packages/reactor/src/executor/util.ts

```diff
@@ -4,14 +4,61 @@ import type {
   DeleteDocumentAction,
   PHDocument,
   UpgradeDocumentAction,
+  UpgradeTransition,
 } from "document-model";
 import { createPresignedHeader, defaultBaseState } from "document-model/core";
+import { DowngradeNotSupportedError } from "../shared/errors.js";
 import type {
   ConsistencyCoordinate,
   ConsistencyToken,
 } from "../shared/types.js";
 import type { OperationWithContext } from "../storage/interfaces.js";
 
+/**
+ * Parses a version string to extract the major version number.
+ * Handles formats like "0", "1", "2", "1.0.0", "2.0", etc.
+ * Returns 0 for empty/invalid strings or "0"/"0.0.0".
+ *
+ * @param versionString - The version string to parse
+ * @returns The extracted major version number
+ */
+export function parseVersionNumber(versionString: string): number {
+  if (!versionString) return 0;
+
+  const parsed = parseInt(versionString, 10);
+  if (!isNaN(parsed)) return parsed;
+
+  const parts = versionString.split(".");
+  if (parts.length > 0) {
+    const major = parseInt(parts[0], 10);
+    if (!isNaN(major)) return major;
+  }
+
+  return 0;
+}
+
+/**
+ * Extracts the target version from an UPGRADE_DOCUMENT action.
+ * First tries action.input.toVersion, then falls back to
+ * initialState.document.version for legacy operations.
+ *
+ * @param action - The UPGRADE_DOCUMENT action
+ * @returns The target version number
+ */
+export function getUpgradeTargetVersion(action: UpgradeDocumentAction): number {
```

Comment:
Same here -- this is `number` now, and it should always be defined or it's an invalid upgrade.

Reply:
added fix in 3bd7f553c3ab3c30e58dad35bedc98baca28d836